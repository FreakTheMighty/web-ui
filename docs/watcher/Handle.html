        <!DOCTYPE html>
        <html>
        <head>
                <meta charset="utf-8">
        <title>Handle&lt;T&gt; class / watcher Library / Dart Documentation</title>
        <link rel="stylesheet" type="text/css"
            href="../styles.css">
        <link href="http://fonts.googleapis.com/css?family=Open+Sans:400,600,700,800" rel="stylesheet" type="text/css">
        <link rel="shortcut icon" href="../favicon.ico">
        
        </head>
        <body data-library="watcher" data-type="Handle&amp;lt;T&amp;gt;">
        <div class="page">
        <div class="header">
          <a href="../index.html"><div class="logo"></div></a>
          <a href="../index.html">Dart Documentation</a>
         &rsaquo; <a href="../watcher.html">watcher</a> &rsaquo; <a href="../watcher/Handle.html">Handle&lt;T&gt;</a>        <div id="search-box">
          <input type="search" name="q" id="q" autocomplete="off"
              class="search-input" placeholder="Search API">
        </div>
        
      </div>
      <div class="drop-down" id="drop-down"></div>
      
        <div class="nav">
        
</div>
<div class="content">
        <h2><strong>Handle&lt;T&gt;</strong>
          class
        </h2>
        
<button id="show-inherited" class="show-inherited">Hide inherited</button>
<div class="doc">
<p>An indirect reference to a value. This is used to create two-way bindings in
MVC applications.</p>
<p>The model can be a normal Dart class. You can then create a handle to a
particular reference so that the view has read/write access without
internally revealing your model abstraction. For example, consider a model
class containing whether or not an item is 'checked' on a list:</p>
<pre><code>class Item {
  int id;
  ...
  bool checked;
</code></pre>
<p>Then we can use a CheckBox view and only reveal the status of the checked
field as follows:</p>
<pre><code>new CheckBoxView(new Handle&lt;bool&gt;(
    () =&gt; item.checked,
    (v) { item.checked = v}));
</code></pre>
<p>A handle with no setter is a read-only handle.</p>
</div>
<div>
<h3>Constructors</h3>
<div class="method"><h4 id="Handle">
<button class="show-code">Code</button>
new <strong>Handle</strong>(<a href="../watcher/Getter.html">Getter</a> _getter, [<a href="../watcher/Setter.html">Setter</a> setter]) <a class="anchor-link" href="#Handle"
              title="Permalink to Handle&lt;T&gt;.Handle">#</a></h4>
<div class="doc">
<p>Create a handle, possibly read-only (if no setter is specified).</p>
<pre class="source">
Handle(this._getter, [Setter&lt;T&gt; setter]) : _setter = setter;
</pre>
</div>
</div>
<div class="method"><h4 id="Handle.of">
<button class="show-code">Code</button>
new <strong>Handle.of</strong>(<a href="../watcher/Handle.html">T</a> value) <a class="anchor-link" href="#Handle.of"
              title="Permalink to Handle&lt;T&gt;.Handle.of">#</a></h4>
<div class="doc">
<pre class="source">
Handle.of(T value) : this(() =&gt; value);
</pre>
</div>
</div>
</div>
<div>
<h3>Methods</h3>
<div class="method inherited"><h4 id="==">
<button class="show-code">Code</button>
<a href="http://api.dartlang.org/dart_core/bool.html">bool</a> <strong>operator ==</strong>(other) <a class="anchor-link" href="#=="
              title="Permalink to Handle&lt;T&gt;.operator ==">#</a></h4>
<div class="inherited-from">inherited from <a href="http://api.dartlang.org/dart_core/Object.html">Object</a> </div><div class="doc">
<p>The equality operator.</p>
<p>The default behavior for all <code>Object</code>s is to return true if and
only if <code>this</code> and 
<span class="param">other</span> are the same object.</p>
<p>If a subclass overrides the equality operator it should override
the <code>hashCode</code> method as well to maintain consistency.</p>
<pre class="source">
bool operator ==(other) =&gt; identical(this, other);
</pre>
</div>
</div>
<div class="method inherited"><h4 id="hashCode">
<button class="show-code">Code</button>
<a href="http://api.dartlang.org/dart_core/int.html">int</a> <strong>hashCode</strong>() <a class="anchor-link" href="#hashCode"
              title="Permalink to Handle&lt;T&gt;.hashCode">#</a></h4>
<div class="inherited-from">inherited from <a href="http://api.dartlang.org/dart_core/Object.html">Object</a> </div><div class="doc">
<p>Get a hash code for this object.</p>
<p>All objects have hash codes. Hash codes are guaranteed to be the
same for objects that are equal when compared using the equality
operator <code>==</code>. Other than that there are no guarantees about
the hash codes. They will not be consistent between runs and
there are no distribution guarantees.</p>
<p>If a subclass overrides <code>hashCode</code> it should override the
equality operator as well to maintain consistency.</p>
<pre class="source">
external int hashCode();
</pre>
</div>
</div>
<div class="method inherited"><h4 id="noSuchMethod">
<button class="show-code">Code</button>
<strong>noSuchMethod</strong>(<a href="http://api.dartlang.org/dart_core/String.html">String</a> name, <a href="http://api.dartlang.org/dart_core/List.html">List</a> args) <a class="anchor-link" href="#noSuchMethod"
              title="Permalink to Handle&lt;T&gt;.noSuchMethod">#</a></h4>
<div class="inherited-from">inherited from <a href="http://api.dartlang.org/dart_core/Object.html">Object</a> </div><div class="doc">
<p><code>noSuchMethod</code> is invoked when users invoke a non-existant method
on an object. The name of the method and the arguments of the
invocation are passed to <code>noSuchMethod</code>. If <code>noSuchMethod</code>
returns a value, that value becomes the result of the original
invocation.</p>
<p>The default behavior of <code>noSuchMethod</code> is to throw a
<code>noSuchMethodError</code>.</p>
<pre class="source">
external Dynamic noSuchMethod(String name, List args);
</pre>
</div>
</div>
<div class="method inherited"><h4 id="runtimeType">
<button class="show-code">Code</button>
<a href="http://api.dartlang.org/dart_core/Type.html">Type</a> <strong>get runtimeType</strong>() <a class="anchor-link" href="#runtimeType"
              title="Permalink to Handle&lt;T&gt;.get runtimeType">#</a></h4>
<div class="inherited-from">inherited from <a href="http://api.dartlang.org/dart_core/Object.html">Object</a> </div><div class="doc">
<p>A representation of the runtime type of the object.</p>
<pre class="source">
external Type get runtimeType;
</pre>
</div>
</div>
<div class="method inherited"><h4 id="toString">
<button class="show-code">Code</button>
<a href="http://api.dartlang.org/dart_core/String.html">String</a> <strong>toString</strong>() <a class="anchor-link" href="#toString"
              title="Permalink to Handle&lt;T&gt;.toString">#</a></h4>
<div class="inherited-from">inherited from <a href="http://api.dartlang.org/dart_core/Object.html">Object</a> </div><div class="doc">
<p>Returns a string representation of this object.</p>
<pre class="source">
external String toString();
</pre>
</div>
</div>
<div class="method"><h4 id="value">
<button class="show-code">Code</button>
<a href="../watcher/Handle.html">T</a> <strong>get value</strong>() <a class="anchor-link" href="#value"
              title="Permalink to Handle&lt;T&gt;.get value">#</a></h4>
<div class="doc">
<pre class="source">
T get value =&gt; _getter();
</pre>
</div>
</div>
<div class="method"><h4 id="value=">
<button class="show-code">Code</button>
void <strong>set value</strong>(<a href="../watcher/Handle.html">T</a> value) <a class="anchor-link" href="#value="
              title="Permalink to Handle&lt;T&gt;.set value">#</a></h4>
<div class="doc">
<pre class="source">
void set value(T value) {
  if (_setter != null) {
    _setter(value);
  } else {
    throw new Exception('Sorry - this handle has no setter.');
  }
}
</pre>
</div>
</div>
</div>
        </div>
        <div class="clear"></div>
        </div>
        
        <div class="footer">
          <div>This page was generated at 2012-10-09 19:04:13.422</div>
        </div>
        <script async src="../client-live-nav.js"></script>
        </body></html>
        
