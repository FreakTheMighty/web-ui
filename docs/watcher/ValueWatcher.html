        <!DOCTYPE html>
        <html>
        <head>
                <meta charset="utf-8">
        <title>ValueWatcher typedef / watcher Library / Dart Documentation</title>
        <link rel="stylesheet" type="text/css"
            href="../styles.css">
        <link href="http://fonts.googleapis.com/css?family=Open+Sans:400,600,700,800" rel="stylesheet" type="text/css">
        <link rel="shortcut icon" href="../favicon.ico">
        
        </head>
        <body data-library="watcher" data-type="ValueWatcher">
        <div class="page">
        <div class="header">
          <a href="../index.html"><div class="logo"></div></a>
          <a href="../index.html">Dart Documentation</a>
         &rsaquo; <a href="../watcher.html">watcher</a> &rsaquo; <a href="../watcher/ValueWatcher.html">ValueWatcher</a>        <div id="search-box">
          <input type="search" name="q" id="q" autocomplete="off"
              class="search-input" placeholder="Search API">
        </div>
        
      </div>
      <div class="drop-down" id="drop-down"></div>
      
        <div class="nav">
        
</div>
<div class="content">
        <h2><strong>ValueWatcher</strong>
          typedef
        </h2>
        
<button id="show-inherited" class="show-inherited">Hide inherited</button>
<div class="doc">
<p>Callback fired when an expression changes.</p>
</div>
<div class="method"><h4 id="ValueWatcher">
<button class="show-code">Code</button>
typedef void ValueWatcher(<a href="../watcher/WatchEvent.html">WatchEvent</a> e) <a class="anchor-link" href="#ValueWatcher"
              title="Permalink to ValueWatcher">#</a></h4>
<div class="doc">
<pre class="source">
typedef void ValueWatcher(WatchEvent e);
</pre>
</div>
</div>
        </div>
        <div class="clear"></div>
        </div>
        
        <div class="footer">
          <div>This page was generated at 2012-10-09 19:04:13.377</div>
        </div>
        <script async src="../client-live-nav.js"></script>
        </body></html>
        
