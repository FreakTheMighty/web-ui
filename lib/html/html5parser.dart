// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Generated by scripts/tree_gen.py.

/**
 * An HTML5 parser written in Dart.
 *
 * Intended to parse all html documents using the [html5 parsing algorithm][p],
 * although the implementation is not yet complete.
 *
 * [p]: http://dev.w3.org/html5/spec/parsing.html#parsing
 */
#library('html5parser');

#source('htmltree.dart');
#source('parser.dart');
#source('tree.dart');
#source('token.dart');
#source('tokenkind.dart');
#source('tokenizer_base.dart');
#source('tokenizer.dart');
