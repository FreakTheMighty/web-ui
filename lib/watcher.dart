// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/**
 * **Deprecated**: import 'web_ui/templating.dart' instead.
 *
 * A library to observe changes on Dart objects.
 *
 * Similar to the principle of watchers in AngularJS, this library provides the
 * mechanisms to observe and react to changes that happen in an application's
 * data model.
 *
 * Watchers have a simple lifetime:
 *
 *   * they are created calling [watch],
 *
 *   * they are fired whenever [dispatch] is called and the watched values
 *   changed since the last time [dispatch] was invoked, and
 *
 *   * they are unregistered using a function that was returned by [watch] when
 *   they were created.
 *
 * For example, you can create a watcher that observes changes to a variable by
 * calling [watch] as follows:
 *
 *     var x = 0;
 *     var stop = watch(() => x, (_) => print('hi'));
 *
 * Changes to the variable 'x' will be detected whenever we call [dispatch]:
 *
 *     x = 12;
 *     x = 13;
 *     dispatch(); // the watcher is invoked ('hi' will be printed once).
 *
 * After deregistering the watcher, events are no longer fired:
 *
 *     stop();
 *     x = 14;
 *     dispatch(); // nothing happens.
 *
 * You can watch several kinds of expressions, including lists. See [watch] for
 * more details.
 *
 * A common design pattern for MVC applications is to call [dispatch] at the end
 * of each event loop (e.g. after each UI event is fired). Our view library does
 * this automatically.
 */
@deprecated library watcher;

import 'package:meta/meta.dart';
export 'package:web_ui/watcher.dart';
