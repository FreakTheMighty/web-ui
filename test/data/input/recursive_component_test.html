<!doctype html>
<!--
   Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
   for details. All rights reserved. Use of this source code is governed by a
   BSD-style license that can be found in the LICENSE file.
-->
<html lang="en">
<head>
  <script type='application/javascript' src="testing.js"></script>
<style> div { padding: 10px; } </style>
</head>
<body>
  Expect nested divs 0 then 1.
  <element name='x-rec' extends="div">
    <template>
      {{check(pos)}}
      <template if="pos < len">
        <div is="x-rec" pos="{{pos + 1}}"></div>
      </template>
    </template>
    <script type='application/dart'>
      import 'package:web_ui/web_ui.dart';
      import 'recursive_test_data.dart';
      class Rec extends WebComponent {
        int pos;
      }
    </script>
  </element>
  <div is="x-rec" pos="{{0}}"></div>

  <script type="application/dart">
  import 'dart:html';
  import 'package:web_ui/watcher.dart';
  import 'recursive_test_data.dart';

  main() {
    window.setTimeout(() {
      dispatch();
      window.setTimeout(() {
        len = 1;
        dispatch();
        window.setTimeout(() {
          len = 3;
          dispatch();
          window.setTimeout(() {
            len = 2;
            dispatch();
            window.setTimeout(() {
              len = 1;
              dispatch();
              window.setTimeout(() => window.postMessage('done', '*'), 0);
            }, 0);
          }, 0);
        }, 0);
      }, 0);
    }, 0);
  }
  </script>
</body>
</html>

