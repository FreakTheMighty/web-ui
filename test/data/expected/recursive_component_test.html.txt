Content-Type: text/plain
layer at (0,0) size 800x600
  RenderView at (0,0) size 800x600
layer at (0,0) size 800x116
  RenderBlock {HTML} at (0,0) size 800x116
    RenderBody {BODY} at (8,8) size 784x100
      RenderBlock (anonymous) at (0,0) size 784x20
        RenderText {#text} at (0,0) size 170x19
          text run at (0,0) width 170: "Expect nested divs 0 then 1."
      RenderBlock {DIV} at (0,20) size 784x80
        RenderBlock (anonymous) at (10,10) size 764x20
          RenderText {#text} at (0,0) size 8x19
            text run at (0,0) width 8: "0"
          RenderText {#text} at (0,0) size 0x0
          RenderText {#text} at (0,0) size 0x0
        RenderBlock {DIV} at (10,30) size 764x40
          RenderText {#text} at (10,10) size 8x19
            text run at (10,10) width 8: "1"
          RenderText {#text} at (0,0) size 0x0
          RenderText {#text} at (0,0) size 0x0
        RenderBlock (anonymous) at (10,70) size 764x0
          RenderText {#text} at (0,0) size 0x0
          RenderText {#text} at (0,0) size 0x0
#EOF
